pkg/v1/tkg/web/src/app/views/landing/wizard/shared/components/steps/metadata-step/metadata-step.component.ts

```diff
@@ -36,48 +46,85 @@ export class MetadataStepComponent extends StepFormDirective implements OnInit {
             const savedLabelsArray = savedLabelsString.split(', ')
             savedLabelsArray.map(label => {
                 const labelArray = label.split(':');
-                this.labels.set(labelArray[0], labelArray[1]);
+                this.addLabel(labelArray[0], labelArray[1]);
             });
         }
         super.initFormWithSavedData();
+        this.formGroup.valueChanges.pipe(
+            distinctUntilChanged((prev, curr) => JSON.stringify(prev) === JSON.stringify(curr)),
+            takeUntil(this.unsubscribe)
+        ).subscribe(() => {
+            this.validateAllLabels();
+        });
     }
 
-    addLabel(key: string, value: string) {
-        if (key === '' || value === '') {
-            this.errorNotification = `Key and value for Labels are required.`;
-        } else if (!this.labels.has(key)) {
-            this.labels.set(key, value);
-            this.formGroup.get('clusterLabels').setValue(this.labels);
-            this.formGroup.controls['newLabelKey'].setValue('');
-            this.formGroup.controls['newLabelValue'].setValue('');
-        } else {
-            this.errorNotification = `A Label with the same key already exists.`;
-        }
+    addLabel(key?: string, value?: string) {
+        const LABEL_KEY_NAME = 'newLabelKey';
+        const LABEL_VALUE_NAME = 'newLabelValue';
+
+        this.labelCounter++;
+        this.labels.set(LABEL_KEY_NAME + this.labelCounter, LABEL_VALUE_NAME + this.labelCounter);
+        this.keySet.add(LABEL_KEY_NAME + this.labelCounter);
+        FormUtils.addControl(
+            this.formGroup,
+            LABEL_KEY_NAME + this.labelCounter,
+            new FormControl(key || '', [
+                this.validationService.isValidLabelOrAnnotation(),
+                this.validationService.isUniqueLabel(
+                    this.formGroup,
+                    this.keySet,
+                    LABEL_KEY_NAME + this.labelCounter)
+            ])
+        );
+
+        FormUtils.addControl(
+            this.formGroup,
+            LABEL_VALUE_NAME + this.labelCounter,
+            new FormControl(value || '', [
+                this.validationService.isValidLabelOrAnnotation()
+            ])
+        );
+        this.validateAllLabels();
+    }
+
+    validateAllLabels () {
+        // The setTimeout could ensure that the validation logic could be ran after
```

Comment:
Nitpick: great documentation, I would just change to:

The setTimeout wrapper ensures that validation logic will run after a new label field is added.

Thanks!

Reply:
fixed